src/v/ssx/include/ssx/single_sharded.h

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/shard_id.hh>
+#include <seastar/core/sharded.hh>
+
+#include <limits>
+#include <optional>
+#include <utility>
+
+namespace ssx {
+
+/// Template helper to run a real service service on one core and an small dummy
+/// on other cores.
+///
+/// \tparam Service a class to be instantiated on each core.  Must expose
+///         a \c stop() method that returns a \c future<>, to be called when
+///         the service is stopped.
+template<typename Service>
+class maybe_service {
+    using underlying_t = std::optional<Service>;
+    underlying_t _underlying;
+
```

Comment:
I'm not understanding why its needed to run a dummy service on another core. It kinda feels like this would be solved using `peering_sharded_service<T>`?

Reply:
> why its needed to run a dummy service on another core

That's to use `ss::sharded<T>` for to run a single service. Sounds weird, but no more than `ss::sharded<T>::start_single`. An alternative idea would be a new wrapper not based on `ss::sharded<T>`, but I anticipate it would require more attention on Seastar upgrade.

I'll also update the PR description with some background.